src/probnum/diffeq/odefiltsmooth/ivpfiltsmooth.py

```diff
@@ -166,175 +162,269 @@ def init_implementation(f, y0, t0, prior, initrv, df=None):
+        we finally compute the Kalman update and return the result. Recall that the error estimate has been computed in the third step.
+        """
+
+        # Read off system matrices; required for calibration / error estimation
+        # Use only where a full call to forward_*() would be too costly.
+        # We use the mathematical symbol `Phi` (and later, `H`), because this makes it easier to read for us.
+        # The system matrix H of the measurement model can be accessed after the first forward_*,
+        # therefore we read it off further below.
+        dt = t_new - t
+        discrete_dynamics = self.dynamics_model.discretise(dt)
+        Phi = discrete_dynamics.state_trans_mat
+
+        # Split the current RV into a deterministic part and a noisy part.
+        # This split is necessary for efficient calibration; see docstring.
+        error_free_state = current_rv.mean.copy()
+        noisy_component = Normal(
+            mean=np.zeros(current_rv.shape),
+            cov=current_rv.cov.copy(),
+            cov_cholesky=current_rv.cov_cholesky.copy(),
         )
 
-        # 3.2 Update the measurement covariance (measure again)
-        meas_rv, info = self.gfilt.measurement_model.forward_rv(
-            rv=pred_rv, t=t_new, compute_gain=True
+        # Compute the measurements for the error-free component
+        pred_rv_error_free, _ = self.dynamics_model.forward_realization(
+            error_free_state, t=t, dt=dt
+        )
+        meas_rv_error_free, _ = self.measurement_model.forward_rv(
+            pred_rv_error_free, t=t
+        )
+        H = self.measurement_model.linearized_model.state_trans_mat_fun(t=t)
+
+        # Compute the measurements for the full components.
+        # Since the means of noise-free and noisy measurements coincide,
+        # we manually update only the covariance.
+        # The first two are only matrix square-roots and will be turned into proper Cholesky factors below.
+        pred_sqrtm = Phi @ noisy_component.cov_cholesky
+        meas_sqrtm = H @ pred_sqrtm
+        full_meas_cov_cholesky = utils.linalg.cholesky_update(
+            meas_rv_error_free.cov_cholesky, meas_sqrtm
+        )
+        full_meas_cov = full_meas_cov_cholesky @ full_meas_cov_cholesky.T
+        meas_rv = Normal(
+            mean=meas_rv_error_free.mean,
+            cov=full_meas_cov,
+            cov_cholesky=full_meas_cov_cholesky,
         )
 
-        # 4. Update
-        zero_data = np.zeros(meas_rv.mean.shape)
-        filt_rv, _ = self.gfilt.measurement_model.backward_realization(
-            zero_data, pred_rv, rv_forwarded=meas_rv, gain=info["gain"]
+        # Estimate local diffusion_model and error
+        diffusion_error_free = self.diffusion_model.calibrate_locally(
+            meas_rv_error_free
+        )
+        diffusion_full = self.diffusion_model.calibrate_locally(meas_rv)
+        diffusion_for_calibration = self.diffusion_model.update_current_information(
```

Comment:
For the `ConstantDiffusion`, in the way as described in Filip's paper, this should just be the identity since there is no calibration, right? If I understand correctly, here you use the current MLE of the global diffusion, and I'm once more not sure we should do that.

Reply:
Ok I'm mistaken. For the local error estimates and adaptive steps, it makes sense that we need to use something here, e.g. the current MLE of the diffusion :) The naming confused me, since I didn't see a "calibration" happening here in that scenario, but that might be on me since I tend to separately think about calibration and error estimation, even though both are closely linked.